File: 41-50/Q42.py
#Q42.py
class Solution:
    def trap(self, h):
        l, r = 0, len(h)-1
        lm = rm = ans = 0
        while l < r:
            if h[l] < h[r]:
                if h[l] >= lm: lm = h[l]
                else: ans += lm - h[l]
                l += 1
            else:
                if h[r] >= rm: rm = h[r]
                else: ans += rm - h[r]
                r -= 1
        return ans

File: 41-50/Q42.cpp
//Q42.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        int l = 0, r = h.size() - 1;
        int lm = 0, rm = 0, ans = 0;
        while(l < r){
            if(h[l] < h[r]){
                if(h[l] >= lm) lm = h[l];
                else ans += lm - h[l];
                l++;
            } else {
                if(h[r] >= rm) rm = h[r];
                else ans += rm - h[r];
                r--;
            }
        }
        return ans;
    }
};
